### PR TITLE
Add j/k to scroll to next/prev tweet

### PR DESCRIPTION
**Pull request by paulmolluzzo**
This is a fix for #10.

I have to say, this feels super hacky to me, but it is working on the timeline and the notifications. There might be some voodoo in there and any feedback is welcome.

### CODE REVIEW

**Comment by paulmolluzzo** [comment]
Ok, I made some improvements because the previous version didn't work at all. I didn't rebase so that you can see the progression of what happened, but the first commit actually doesn't work at all and the new version is working even as I go 20-40 tweets down in the timeline.

I am betting that there's going to be a lot of feedback on the code, and I'm cool with that. Please let me know if you find bugs or have code style feedback.

File: extension/content.js [hunk 1]
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
...
+	Mousetrap.bind('k', event => {
+		scrollToTweet(event);
+	});
+}
+
+function scrollToTweet(event) {
+	const navHeight = $('nav').clientHeight;
+	const tweets = document.querySelectorAll('._222QxFjc[role="row"]');
+	const currentTop = window.scrollY;
+	const keyCode = event.charCode;
```
Comment (sindresorhus) [inline extension/content.js]: Can you add a `TODO` comment here about using https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/key when Chrome 51 is out?

File: extension/content.js [hunk 2]
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
 			$('button[data-testid="dmComposerSendButton"]').click();
 		}
 	});
+
+	Mousetrap.bind('j', event => {
+		scrollToTweet(event);
+	});
```
Comment (sindresorhus) [inline extension/content.js]:
```js
Mousetrap.bind('j', scrollToTweet);
```

File: extension/content.js [hunk 3]
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
...
+	const tweets = document.querySelectorAll('._222QxFjc[role="row"]');
+	const currentTop = window.scrollY;
+	const keyCode = event.charCode;
+	// this padding has to be applied to factor in the padding that's added to the container
+	const tweetsContainerPadding = Math.ceil(parseInt($('._1nQuzuNK').style.paddingTop.replace('px', ''), 10));
+
+	// shortcut for getting the total offset from the top of the document of a particular tweet
+	const totalOffset = tweetOffset => {
+		return tweetOffset + navHeight + tweetsContainerPadding;
+	};
```
Comment (sindresorhus) [inline extension/content.js]:
```js
const totalOffset = tweetOffset => tweetOffset + navHeight + tweetsContainerPadding;
```

File: extension/content.js [hunk 4]
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
...
+
+	// takes a tweet's offsetTop and checks if it's below the navigation bar
+	const tweetIsBelowNav = offset => {
+		return totalOffset(offset) > currentTop;
+	};
```
Comment (sindresorhus) [inline extension/content.js]: Same as https://github.com/sindresorhus/refined-twitter/pull/19/files#r62426180

File: extension/content.js [hunk 5]
```diff
@@ -54,6 +54,56 @@ function registerShortcuts(username) {
...
+
+	let scrollTarget = 0;
+
+	Array.from(tweets).some((tweet, index) => {
```
Comment (sindresorhus) [inline extension/content.js]: Can you add a `TODO` about removing `Array.from` when Chrome 51 is a target? With that version DOM collections are iterable \o/ https://mobile.twitter.com/chromiumdev/status/717736215433256960

**Comment by sindresorhus** [comment]
This is awesome! Works almost perfectly. I only noticed one problem. On some tweets, I have to press `j` twice to get to the next tweet:

untitled12

**Comment by sindresorhus** [comment]
Could you add the `j`/`k` shortcuts to the readme?

**Comment by sindresorhus** [comment]
Maybe put it in a separate file so it can easily be ported to [Anatine](https://github.com/sindresorhus/anatine)?

**Comment by paulmolluzzo** [comment]
@sindresorhus OK, I've addressed all the items you mentioned. Specifically:

* Fixed that occassional double-press scenario (see below)
* Added this to the README
* Streamlined the `totalOffset` and `tweetIsBelowNav` functions
* Added TODO items
* Moved this all to a new file
* Squashed all commits

Some notes:

The double-press issue is because of the way the math works out with all the `offset` grabbing and rounding, and in the end a magic number fixed it the best (I tried switching `ceil` to `floor` and changing `offset` to `getBoundingClientRect().top` and none of that worked for me). I'm not crazy about it, but it works. You can actually see the issue if you look - sometimes the top border of tweet perfect lines up, sometimes it's off by 1-5px.

I'm ignoring the first line of the new file to avoid the `no-unused-vars` error from xo.

**Comment by sindresorhus** [comment]
Soooo good! :)

**Comment by sindresorhus** [comment]
I ported it to Anatine: https://github.com/sindresorhus/anatine/commit/b70b860f1e86ab640b2e7e644e5c3efdded385ab :)

The focused tweet is not completely placed at the top though. Would you mind taking a look?

napkin 09-05-16 2 25 04 pm

**Comment by paulmolluzzo** [comment]
@sindresorhus yup, I'll take a look.

**Comment by paulmolluzzo** [comment]
@sindresorhus If it's working properly for refined-twitter, I suggest merging this PR and then handling the bug separately for anatine. It's possible that the solution needs to be customized for each repo based on the selectors/DOM we have in each.

**Comment by sindresorhus** [comment]
@paulmolluzzo This is already merged.

**Comment by hkdobrev** [comment]
I admit I still haven't installed the extension. But I've used Twitter.com for years and `j/k` navigation is provided by Twitter itself for a really long time. I'm sure it was available around 2009-2010 and it is still available now without the extension. So what's this about?

**Comment by hkdobrev** [comment]
I'm sorry for commenting unprepared. So mobile.twitter.com doesn't have the keyboard shortcuts and this extension is all about mobile.twitter.com on desktop. This change makes perfect sense now. I'm probably not going to use it though. 😸

**Comment by paulmolluzzo** [comment]
@hkdobrev it's already merged anyway :)

**Pull request closed without merging**